src/lib.rs

```diff
@@ -95,161 +95,121 @@ impl Slice {
-                Backwards => ensure_within(n, 1..len + 1).map(|n| len - n),
-            },
+            &Head(n) => Some(n as i128),
+            &Tail(n) => Some(-(n as i128)),
             Default => None,
         }
     }
 }
 
-/// Return Some(n) if r.start <= n < r.end, otherwise return None.
-fn ensure_within(n: usize, r: Range<usize>) -> Option<usize> {
-    if r.contains(&n) {
-        Some(n)
-    } else {
-        None
-    }
+/// A SSlice is a slice implemented using signed arithmetics. It uses a signed type that
+/// must be able to represent usize. This impl won't work if size_of<usize> >= 16.
+/// 64 bits of address space should be enough for everybody.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct SSlice {
+    start: Option<i128>,
+    end: Option<i128>,
+    step: Option<i128>,
 }
 
-/// An iterator that counts from an initial number up to a (included or excluded) final limit.
-/// The direction and stride of the iteration can be controlled by the step parameter.
-/// A zero step produces an empty iteration.
-#[derive(Debug)]
-struct RangeIterator {
-    end: Bound<usize>,
-    step: isize,
-    pos: usize,
-    done: bool,
-    dir: Direction,
-}
+impl SSlice {
+    /// Returns an iterator that yields the indices that match the slice expression.
+    fn indices(&self, ulen: usize) -> impl Iterator<Item = usize> {
+        let len = ulen as i128;
 
-impl RangeIterator {
-    fn new(start: usize, end: Bound<usize>, step: isize) -> Self {
-        RangeIterator {
-            pos: start,
-            end,
-            step,
-            done: false,
-            dir: Direction::new(step),
+        if len == 0 {
+            return Iter {
+                i: 0,
+                end: 0,
+                step: 0,
+            };
         }
-    }
-}
 
-// Iteration direction, knows how to compare the limit index.
-#[derive(Debug)]
-enum Direction {
-    Forwards,
-    Backwards,
-}
+        let step = self.step.unwrap_or(1);
 
-impl Direction {
-    fn new(step: isize) -> Self {
-        if step >= 0 {
-            Direction::Forwards
+        let (def_start, def_end) = if step >= 0 {
+            (0, len)
         } else {
-            Direction::Backwards
+            (len - 1, -1)
+        };
+
+        let (min, max) = if step >= 0 {
+            (def_start, def_end)
+        } else {
+            // when iterating backwards, start becomes the higher bound
+            // and end becomes the lower bound. In order to iterate
+            // down to the 0th index, we must allow -1 as the non
+            // inclusive lower bound (in the same way as we allow "len"
+            // as the upper bound when iterating forward).
+            (def_end, def_start)
+        };
+
+        // negative slice parameters are relative to the end of the array.
+        let to_index = |n: i128| if n >= 0 { n } else { len + n };
+
+        Iter {
+            i: clamp(self.start.map(to_index).unwrap_or(def_start), min..=max),
```

Comment:
@glyn I rearranged the "to_index" to happen before applying the defaults, which means that now the defaults
can be expressed as values in the "array index domain" as opposed to the "slice bound parameter domain" (I'd like to capture that distinction in the type system, TODO another time). This makes it possible to use more intuitive default values and also reuse the same values for the min and max bounds.

Reply:
That's what I hoping for. Glad it's working out.